src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java

```diff
@@ -246,84 +360,91 @@ public void resetNeededResources()
+            builderInventory = builder.getInventoryCitizen();
         }
 
-        /**
-         * Gets the blockOut Window.
-         *
-         * @return the window of the builder building.
-         */
-        @NotNull
-        @Override
-        public Window getWindow()
-        {
-            return new WindowHutBuilder(this);
-        }
 
-        @Override
-        public void deserialize(@NotNull ByteBuf buf)
+        for (@NotNull final Map.Entry<String, ItemStack> entry : neededResources.entrySet())
         {
-            super.deserialize(buf);
+            final ItemStack itemStack = entry.getValue();
+            BuildingBuilderResource resource = resourcesAvailable.get(entry.getKey());
+            if (resource == null)
+            {
+                resource = new BuildingBuilderResource(entry.getValue().getDisplayName(), itemStack,0,itemStack.getCount());
+            }
 
-            final int size = buf.readInt();
-            neededResources = new HashMap<>();
+            if (builderInventory!=null)
+            {
+                resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory(builderInventory, entry.getValue().getItem(), entry.getValue().getItemDamage()));
+            }
 
-            for (int i = 0; i < size; i++)
+            final IInventory chestInventory = this.getTileEntity();
+            if (chestInventory!=null)
             {
-                final String block = ByteBufUtils.readUTF8String(buf);
-                final int amount = buf.readInt();
-                neededResources.put(block, amount);
+                resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory(chestInventory, entry.getValue().getItem(), entry.getValue().getItemDamage()));
             }
-        }
 
-        @NotNull
-        @Override
-        public Skill getPrimarySkill()
-        {
-            return Skill.INTELLIGENCE;
+            //Count in the additional chests as well
+            if (builder!=null)
+            {
+                for(final BlockPos pos : getAdditionalCountainers())
+                {
+                    final TileEntity entity = builder.world.getTileEntity(pos);
+                    if(entity instanceof TileEntityChest)
+                    {
+                        resource.setAvailable(resource.getAvailable() 
+                            + InventoryUtils.getItemCountInInventory((TileEntityChest)entity, entry.getValue().getItem(), entry.getValue().getItemDamage()));
+                    }
+                }
+            }
+
+            resourcesAvailable.put(entry.getKey(), resource);
         }
+    }
 
-        @NotNull
-        @Override
-        public Skill getSecondarySkill()
+    /*
+     * Check if the builder requires a certain ItemStack for the current construction.
+     *
+     * @param stack the stack to test.
+     * @return true if so.
+     */
+    public boolean requiresResourceForBuilding(ItemStack stack)
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)